src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectFactory.cs

```diff
@@ -23,33 +28,54 @@ internal sealed class VisualStudioProjectFactory : IVsTypeScriptVisualStudioProj
     {
         private const string SolutionContextName = "Solution";
         private const string SolutionSessionIdPropertyName = "SolutionSessionID";
-
+        private readonly IThreadingContext _threadingContext;
         private readonly VisualStudioWorkspaceImpl _visualStudioWorkspaceImpl;
         private readonly HostDiagnosticUpdateSource _hostDiagnosticUpdateSource;
+        private readonly Shell.IAsyncServiceProvider _serviceProvider;
         private readonly ImmutableArray<Lazy<IDynamicFileInfoProvider, FileExtensionsMetadata>> _dynamicFileInfoProviders;
 
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
         public VisualStudioProjectFactory(
+            IThreadingContext threadingContext,
             VisualStudioWorkspaceImpl visualStudioWorkspaceImpl,
             [ImportMany] IEnumerable<Lazy<IDynamicFileInfoProvider, FileExtensionsMetadata>> fileInfoProviders,
-            HostDiagnosticUpdateSource hostDiagnosticUpdateSource)
+            HostDiagnosticUpdateSource hostDiagnosticUpdateSource,
+            SVsServiceProvider serviceProvider)
         {
+            _threadingContext = threadingContext;
             _visualStudioWorkspaceImpl = visualStudioWorkspaceImpl;
             _dynamicFileInfoProviders = fileInfoProviders.AsImmutableOrEmpty();
             _hostDiagnosticUpdateSource = hostDiagnosticUpdateSource;
+            _serviceProvider = (Shell.IAsyncServiceProvider)serviceProvider;
         }
 
-        public VisualStudioProject CreateAndAddToWorkspace(string projectSystemName, string language)
-            => CreateAndAddToWorkspace(projectSystemName, language, new VisualStudioProjectCreationInfo());
+        public Task<VisualStudioProject> CreateAndAddToWorkspaceAsync(string projectSystemName, string language, CancellationToken cancellationToken)
+            => CreateAndAddToWorkspaceAsync(projectSystemName, language, new VisualStudioProjectCreationInfo(), cancellationToken);
 
-        public VisualStudioProject CreateAndAddToWorkspace(string projectSystemName, string language, VisualStudioProjectCreationInfo creationInfo)
+        public async Task<VisualStudioProject> CreateAndAddToWorkspaceAsync(
+            string projectSystemName, string language, VisualStudioProjectCreationInfo creationInfo, CancellationToken cancellationToken)
         {
-            // HACK: Fetch this service to ensure it's still created on the UI thread; once this is moved off we'll need to fix up it's constructor to be free-threaded.
+            // HACK: Fetch this service to ensure it's still created on the UI thread; once this is
+            // moved off we'll need to fix up it's constructor to be free-threaded.
+
+            // Since we're on the UI thread here anyways, use that as an opportunity to grab the
+            // IVsSolution object and solution file path.
```

Comment:
Comment should be moved to the IVsSolution-interacting stuff.

Reply:
done.